File: day15.fs
namespace AOC

module Day15 =

    let rec part1 (m: Map<int, int>) age prevNumber =
        match age with
            | 30000000 -> prevNumber
            | _ ->
                match m.TryFind(prevNumber) with
                    | None -> part1 (m.Add (prevNumber, age)) (age + 1) 0
                    | Some(n) ->
                        let prev = age - n
                        let map = m.Add (prevNumber, age)
                        part1 map (age + 1) prev

    [<EntryPoint>]
    let main args =
        let startMap = Map.ofList [1, 1;
                                   20, 2;
                                   8, 3;
                                   12, 4;
                                   0, 5;]
        let n = part1 startMap 6 14
        printfn "2020th: %i" n
        0

File: day15.hs
module Main where

import qualified Data.Map.Strict as M

part1 :: M.Map Int Int -> Int -> Int -> Int
part1 m age prevNumber =
  case age of
    30000000 -> prevNumber
    _ -> case M.lookup prevNumber m of
      Nothing -> part1 (M.insert prevNumber age m) (1 + age) 0
      Just(n) -> let prev = age - n
                     map = M.insert prevNumber age m
                 in part1 map (1 + age) prev

main :: IO ()
main = do
  let startMap = M.fromList [(1,1), (20, 2), (8, 3), (12, 4), (0,5)]
  let n = part1 startMap 6 14
  print n
  return ()
